src/CodeGen_LLVM.cpp

```diff
@@ -1910,63 +1910,76 @@ void CodeGen_LLVM::visit(const Load *op) {
-            // from the load, true indicates we should take the
-            // odd-numbered-lanes.
-            bool shifted_a = false, shifted_b = false;
+        } else if (ramp && stride && 2 <= stride->value && stride->value <= 4) {
+            // Try to rewrite strided loads as shuffles of dense loads,
+            // aligned to the stride. This makes adjacent strided loads
+            // share the same underlying dense loads.
+            ModulusRemainder align = op->alignment;
+            Expr base = ramp->base;
+            int aligned_stride = gcd(stride->value, align.modulus);
+            int offset = 0;
+            if (aligned_stride == stride->value) {
+                offset = mod_imp((int)align.remainder, aligned_stride);
+            } else {
+                const Add *add = base.as<Add>();
+                if (const IntImm *add_c = add ? add->b.as<IntImm>() : base.as<IntImm>()) {
+                    offset = mod_imp(add_c->value, stride->value);
+                }
+            }
 
-            bool external = op->param.defined() || op->image.defined();
+            if (offset) {
+                base = simplify(base - offset);
+                align.remainder -= offset;
+            }
 
-            // Don't read beyond the end of an external buffer.
+            // We want to load a few more bytes than the original load did.
+            // This is safe under these conditions:
+            // - The alignment information is sufficient to determine the
+            //   modified load is safe.
+            // - For internal buffers, we know this is safe because we
+            //   allocate an extra 3 values of padding.
             // (In ASAN mode, don't read beyond the end of internal buffers either,
             // as ASAN will complain even about harmless stack overreads.)
-            if (external || target.has_feature(Target::ASAN)) {
-                base_b -= 1;
-                align_b = align_b - 1;
-                shifted_b = true;
-            } else {
-                // If the base ends in an odd constant, then subtract one
-                // and do a different shuffle. This helps expressions like
-                // (f(2*x) + f(2*x+1) share loads
-                const Add *add = ramp->base.as<Add>();
-                const IntImm *offset = add ? add->b.as<IntImm>() : ramp->base.as<IntImm>();
-                if (offset && offset->value & 1) {
-                    base_a -= 1;
-                    align_a = align_a - 1;
-                    shifted_a = true;
-                    base_b -= 1;
-                    align_b = align_b - 1;
-                    shifted_b = true;
-                }
+            // The min moves lower by offset.
+            int load_lanes = ramp->lanes * stride->value;
+            bool alignment_safe = align.remainder >= 0;
+            // The max needs to not cross an alignment boundary.
+            if (align.modulus > 0) {
```

Comment:
I'm confused about this check. align.modulus could be arbitrarily large and is not necessarily a power of two, so I don't see why this makes it safe. Should this be a check on aligned_stride instead?

Reply:
I have been unsure about this check in general, especially without #5793, this alignment doesn't include the host alignment.

I'm also checking and performance without the check at all actually seems fine, I think I'll just remove it.